File: cs50x-2022/week2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

// Function Prototypes
short IndexCalculator(string);

int main(void)
{
    // Receive a text from the user 
    string text = get_string("Text: ");

    // Calculates the Coleman-Liau index
    short index = IndexCalculator(text);
    
    // Print the index
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);    
    }
}

short IndexCalculator(string str)
{
    unsigned int l = 0, w = 1, s = 0;
    for (int i = 0, x = strlen(str); i < x; i++)
    {
        // Counts letters
        if ((str[i] >= 'a' && str[i] <= 'z') || (str[i] >= 'A' && str[i] <= 'Z'))
        {
            l++;
        }
        
        // Counts words
        if (str[i] == ' ')
        {
            w++;
        }
        
        // Counts sentences 
        if (str[i] == '.' || str[i] == '?'  || str[i] == '!')
        {
            s++;
        }
    }
    
    // Calculate the Coleman-Liau index
    float L = l * (100.0 / w);
    float S = s * (100.0 / w);
    short index = round(0.0588 * L - 0.296 * S - 15.8);
    
    return index;
}

File: cs50x-2022/week6/sentimental-readability/readability.py
from cs50 import get_string


def main():
    # Receives a text from the user
    text = get_string("Text: ")

    # Gets the Coleman-Liau index
    index = index_calculator(text)

    # Prints the index
    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


def index_calculator(text):
    letters = 0
    words = 0
    sentences = 0
    last_char = ""

    char_list = [",", ":", ";", "\"", " ", ".", "?", "!"]

    for ch in text:
        # Counts letters
        if ch.isalpha():
            letters += 1
        # Counts words
        if ch in char_list and last_char.isalpha():
            words += 1
        # Counts sentences
        if ch in char_list[5:]:
            sentences += 1

        last_char = ch

    #  Calculates and returns the Coleman-Liau index
    L = letters * (100.0 / words)
    S = sentences * (100.0 / words)

    return round((0.0588 * L) - (0.296 * S) - 15.8)


if __name__ == "__main__":
    main()